<a href="https://colab.research.google.com/github/vector777-ai/DeepSpaceExoplanetHunting/blob/master/LunarRockClassifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# Load the Drive helper and mount
from google.colab import drive

# This will prompt for authorization.
drive.mount('/content/drive',force_remount=True)

Mounted at /content/drive

In [0]:
cd drive

/content/drive

In [0]:
cd My Drive

/content/drive/My Drive

In [0]:
cd LunarRockDataSet

/content/drive/My Drive/LunarRockDataSet

In [0]:
import os
import shutil
import numpy as np
import pandas as pd
import tensorflow as tf
import keras
from keras import backend as K
from keras.preprocessing.image import ImageDataGenerator
from keras import layers
from keras.layers import Input,Dense,Activation,ZeroPadding2D,BatchNormalization,Flatten,Conv2D
from keras.layers import MaxPooling2D,Dropout,Lambda,Concatenate
from keras.models import Model,Sequential
from keras.optimizers import Adam
K.clear_session()

Using TensorFlow backend.

In [0]:
train_minibatch_size=20
dev_minibatch_size=10
test_minibatch_size=25
input_shape=(480,720,3)

In [0]:
labels=["Large","Small"]
train_batches=ImageDataGenerator(rescale=1./225).flow_from_directory(
            directory='/content/drive/My Drive/LunarRockDataSet/TrainSet',
            target_size=(480,720),
            color_mode='rgb',
            classes=labels,
            batch_size=train_minibatch_size,
            class_mode='categorical',
            shuffle=True)

Found 7237 images belonging to 2 classes.

In [0]:
dev_batches=ImageDataGenerator(rescale=1./225).flow_from_directory(
            directory='/content/drive/My Drive/LunarRockDataSet/DevSet',
            target_size=(480,720),
            color_mode='rgb',
            classes=labels,
            batch_size=dev_minibatch_size,
            class_mode='categorical',
            shuffle=True)

Found 2412 images belonging to 2 classes.

In [0]:
test_batches=ImageDataGenerator(rescale=1./225).flow_from_directory(
            directory='/content/drive/My Drive/LunarRockDataSet/TestSet',
            target_size=(480,720),
            color_mode='rgb',
            classes=labels,
            batch_size=test_minibatch_size,
            class_mode='categorical',
            shuffle=True)

Found 2410 images belonging to 2 classes.

In [0]:
model = Sequential(name='forward_prop')
model.add(Conv2D(32,kernel_size=(3,3),strides=(1,1),padding='same',input_shape=input_shape,name='conv1'))
model.add(Conv2D(32,kernel_size=(3,3),strides=(1,1),padding='same',input_shape=input_shape,name='conv1.1'))
model.add(MaxPooling2D((3,3),strides=2))
model.add(Conv2D(64,kernel_size=(3,3),strides=(1,1),padding='same',name='conv2'))
model.add(Conv2D(64,kernel_size=(3,3),strides=(1,1),padding='same',name='conv2.1'))
model.add(MaxPooling2D((5,5),strides=2))
model.add(Conv2D(128,kernel_size=(3,3),strides=(1,1),padding='same',name='conv3'))
model.add(Conv2D(128,kernel_size=(3,3),strides=(1,1),padding='same',name='conv3.1'))
model.add(Conv2D(128,kernel_size=(3,3),strides=(1,1),padding='same',name='conv3.2'))
model.add(MaxPooling2D((7,7),strides=2))
model.add(Conv2D(128,kernel_size=(3,3),strides=(1,1),padding='same',name='conv4'))
model.add(Conv2D(128,kernel_size=(3,3),strides=(1,1),padding='same',name='conv4.1'))
model.add(MaxPooling2D((3,3),strides=2))
model.add(Conv2D(256,kernel_size=(3,3),strides=(1,1),padding='same',name='conv5'))
model.add(Conv2D(256,kernel_size=(3,3),strides=(1,1),padding='same',name='conv5.1'))
model.add(MaxPooling2D((3,3),strides=2))
model.add(Conv2D(512,kernel_size=(3,3),strides=(1,1),padding='same',name='conv6'))
model.add(MaxPooling2D((3,3),strides=2))
model.add(Flatten())
model.add(Dense(4096,activation='relu',name='fc0'))
model.add(Dense(4096,activation='relu',name='fc1'))
model.add(Dense(2,activation='sigmoid',name='fc2'))
model.compile(Adam(lr=.0001),'categorical_crossentropy',metrics=['accuracy'])

In [0]:
model.summary()

Model: "forward_prop"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1 (Conv2D)               (None, 480, 720, 32)      896       
_________________________________________________________________
conv1.1 (Conv2D)             (None, 480, 720, 32)      9248      
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 239, 359, 32)      0         
_________________________________________________________________
conv2 (Conv2D)               (None, 239, 359, 64)      18496     
_________________________________________________________________
conv2.1 (Conv2D)             (None, 239, 359, 64)      36928     
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 118, 178, 64)      0         
_________________________________________________________________
conv3 (Conv2D)               (None, 118, 178, 128)    

In [0]:
model.fit_generator(train_batches, steps_per_epoch=20, validation_data=dev_batches, validation_steps=20, epochs=15, verbose=1)

Epoch 1/15
20/20 [==============================] - 801s 40s/step - loss: 0.5577 - acc: 0.7125 - val_loss: 0.5466 - val_acc: 0.8000
Epoch 2/15
20/20 [==============================] - 784s 39s/step - loss: 0.4350 - acc: 0.8250 - val_loss: 0.2656 - val_acc: 0.9300
Epoch 3/15
20/20 [==============================] - 782s 39s/step - loss: 0.3021 - acc: 0.8925 - val_loss: 0.1709 - val_acc: 0.9450
Epoch 4/15
20/20 [==============================] - 814s 41s/step - loss: 0.3066 - acc: 0.9075 - val_loss: 0.2533 - val_acc: 0.8850
Epoch 5/15
20/20 [==============================] - 813s 41s/step - loss: 0.2098 - acc: 0.9225 - val_loss: 0.1529 - val_acc: 0.9250
Epoch 6/15
20/20 [==============================] - 792s 40s/step - loss: 0.1190 - acc: 0.9600 - val_loss: 0.0670 - val_acc: 0.9850
Epoch 7/15
20/20 [==============================] - 782s 39s/step - loss: 0.0783 - acc: 0.9750 - val_loss: 0.0546 - val_acc: 0.9800
Epoch 8/15
20/20 [==============================] - 800s 40s/step - loss: 0.

Epoch 1/15
20/20 [==============================] - 801s 40s/step - loss: 0.5577 - acc: 0.7125 - val_loss: 0.5466 - val_acc: 0.8000

Epoch 2/15
20/20 [==============================] - 784s 39s/step - loss: 0.4350 - acc: 0.8250 - val_loss: 0.2656 - val_acc: 0.9300

Epoch 3/15
20/20 [==============================] - 782s 39s/step - loss: 0.3021 - acc: 0.8925 - val_loss: 0.1709 - val_acc: 0.9450

Epoch 4/15
20/20 [==============================] - 814s 41s/step - loss: 0.3066 - acc: 0.9075 - val_loss: 0.2533 - val_acc: 0.8850

Epoch 5/15
20/20 [==============================] - 813s 41s/step - loss: 0.2098 - acc: 0.9225 - val_loss: 0.1529 - val_acc: 0.9250

Epoch 6/15
20/20 [==============================] - 792s 40s/step - loss: 0.1190 - acc: 0.9600 - val_loss: 0.0670 - val_acc: 0.9850

Epoch 7/15
20/20 [==============================] - 782s 39s/step - loss: 0.0783 - acc: 0.9750 - val_loss: 0.0546 - val_acc: 0.9800

Epoch 8/15
20/20 [==============================] - 800s 40s/step - loss: 0.1564 - acc: 0.9550 - val_loss: 0.0747 - val_acc: 0.9850

Epoch 9/15
20/20 [==============================] - 800s 40s/step - loss: 0.0449 - acc: 0.9875 - val_loss: 0.1245 - val_acc: 0.9750

Epoch 10/15
20/20 [==============================] - 798s 40s/step - loss: 0.0149 - acc: 0.9925 - val_loss: 0.0202 - val_acc: 0.9950

Epoch 11/15
20/20 [==============================] - 783s 39s/step - loss: 0.0315 - acc: 0.9875 - val_loss: 0.0065 - val_acc: 1.0000

Epoch 12/15
20/20 [==============================] - 784s 39s/step - loss: 0.1256 - acc: 0.9750 - val_loss: 0.1363 - val_acc: 0.9350

Epoch 13/15
20/20 [==============================] - 784s 39s/step - loss: 0.1248 - acc: 0.9750 - val_loss: 0.0149 - val_acc: 1.0000

Epoch 14/15
20/20 [==============================] - 778s 39s/step - loss: 0.0273 - acc: 0.9925 - val_loss: 0.0583 - val_acc: 0.9900

Epoch 15/15
20/20 [==============================] - 777s 39s/step - loss: 0.0172 - acc: 0.9925 - val_loss: 0.0206 - val_acc: 0.9850


<keras.callbacks.History at 0x7f7cbeed27f0>

In [0]:
train_acc = model.evaluate_generator(train_batches, steps=1, verbose=0)
print("Train Accuracy: "+str(train_acc[1]))

Train Accuracy: 1.0

In [0]:
dev_acc = model.evaluate_generator(dev_batches, steps=1, verbose=0)
print("Dev Accuracy: "+str(dev_acc[1]))

Dev Accuracy: 1.0

In [0]:
test_acc = model.evaluate_generator(test_batches, steps=1, verbose=0)
print("Test Accuracy: "+str(test_acc[1]))

Test Accuracy: 0.9599999785423279

In [0]:
model_json = model.to_json()
with open("model.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model.save_weights("model.h5")
print("Saved model to disk")

Saved model to disk